File: src/app/core/ot/Protocol.ts

export type ResponseHandler<A> = (m: Response<A>) => void;
export interface Connection<A> {
    Propose(p: ProposeEntry<A>): void;
    OnMessage(h: ResponseHandler<A>): void;
}

//
// Message sent TO the client
//

export interface AcceptedEntry<A> {
    Action: A;
    ProposedClock: number;
    AcceptedClock: number;
    SchemaVersion: string;
    UserID: string;
    SessionID: string;
}

export function MapAE<A, B>(a: AcceptedEntry<A>, f: (_: A) => B): AcceptedEntry<B> {
    return {
        Action: f(a.Action),
        ProposedClock: a.ProposedClock,
        AcceptedClock: a.AcceptedClock,
        SchemaVersion: a.SchemaVersion,
        UserID: a.UserID,
        SessionID: a.SessionID,
    };
}


export interface ProposeAck {
    kind: "ProposeAck";
    AcceptedClock: number;
}

export interface WelcomeMessage<A> {
    kind: "WelcomeMessage";
    SessionID: string;
    MissedEntries: AcceptedEntry<A>[];
}

export function MapWM<A, B>(a: WelcomeMessage<A>, f: (_: A) => B): WelcomeMessage<B> {
    return {
        kind: a.kind,
        SessionID: a.SessionID,
        MissedEntries: a.MissedEntries.map(v => MapAE(v, f)),
    };
}

export interface NewEntries<A> {
    kind: "NewEntries";
    Entries: AcceptedEntry<A>[];
}

export function MapNE<A, B>(a: NewEntries<A>, f: (_: A) => B): NewEntries<B> {
    return {
        kind: a.kind,
        Entries: a.Entries.map(v => MapAE(v, f)),
    };
}

export interface CloseMessage {
    kind: "CloseMessage";
    Reason: string;
}

export type Response<A> =
    | ProposeAck
    | WelcomeMessage<A>
    | NewEntries<A>
    | CloseMessage;


//
// Message sent FROM the client
//

export interface ProposeEntry<A> {
    kind: "ProposeEntry"
    Action: A;
    ProposedClock: number;
    SchemaVersion: string;
}

export function MapPE<A, B>(a: ProposeEntry<A>, f: (_: A) => B): ProposeEntry<B> {
    return {
        kind: "ProposeEntry",
        Action: f(a.Action),
        ProposedClock: a.ProposedClock,
        SchemaVersion: a.SchemaVersion,
    }
}

export interface JoinDocument {
    kind: "JoinDocument"
    LogClock: number;
    AuthType: string;
    AuthID: string;
}

export type Message<A> =
    | ProposeEntry<A>
    | JoinDocument;

File: src/server/ot/conn/protocol.go
package conn

import (
	"encoding/json"
	"errors"
	"fmt"

	"github.com/OpenCircuits/OpenCircuits/site/go/core/model"
	"github.com/OpenCircuits/OpenCircuits/site/go/ot/doc"
)

//
// Types in this file are mirrored in ot/Protocol.ts
//

type RawMessageWrapper struct {
	Type string
	Msg  json.RawMessage
}

//
// Message sent TO the client
//

type (
	ProposeAck   = doc.ProposeAck
	CloseMessage = doc.CloseMessage
)

type WelcomeMessage struct {
	SessionID     model.SessionID
	MissedEntries []doc.AcceptedEntry
}

type NewEntries struct {
	Entries []doc.AcceptedEntry
}

func Serialize(s interface{}) ([]byte, error) {
	var t string
	switch s.(type) {
	case ProposeAck:
		t = "ProposeAck"
	case WelcomeMessage:
		t = "WelcomeMessage"
	case NewEntries:
		t = "NewEntries"
	case CloseMessage:
		t = "CloseMessage"
	default:
		return nil, errors.New("serialized invalid type (INTERNAL PROTOCOL VIOLATION)")
	}

	rawMsg, err := json.Marshal(s)
	if err != nil {
		return nil, err
	}
	return json.Marshal(RawMessageWrapper{t, rawMsg})
}

//
// Messages sent FROM the client
//

type ProposeEntry struct {
	Action        json.RawMessage
	ProposedClock uint64
	SchemaVersion model.SchemaVersion
}

type JoinDocument struct {
	LogClock uint64
	AuthType string
	AuthID   string
}

func Deserialize(data []byte) (interface{}, error) {
	var rawMsg RawMessageWrapper
	err := json.Unmarshal(data, &rawMsg)
	if err != nil {
		return nil, err
	}

	t := rawMsg.Type
	switch {
	case t == "ProposeEntry":
		var m ProposeEntry
		err := json.Unmarshal(rawMsg.Msg, &m)
		return m, err
	case t == "JoinDocument":
		var m JoinDocument
		err := json.Unmarshal(rawMsg.Msg, &m)
		return m, err
	default:
		return nil, errors.New(fmt.Sprint("received invalid message kind: ", rawMsg))
	}
}
